bitcoin/src/blockdata/script.rs

```diff
@@ -400,6 +400,15 @@ impl Script {
       Builder::new()
     }
 
+    /// Returns the public key if this script is P2PK with a valid public key.
+    pub fn p2pk_public_key(&self) -> Option<PublicKey> {
+        if self.is_p2pk() {
+            PublicKey::from_slice(&self.0[1..66]).ok()
```

Comment:
Oh, actually I just checked the implementation of `is_p2pk` and it also considers compressed keys but this doesn't. So definitely it needs to check the length.

But this is duplicating the code so I wonder if we should implement whole logic here and then call `p2pk_public_key().is_some()` in `is_p2pk()`. However we would stop returning `true` on invalid keys which seems wrong.

Perhaps:

```rust
// maybe should be pub?
fn p2pk_pubkey_bytes(&self) -> Option<&[u8]> {
   match self.len() {
        // ...
    }
}
```

And then call it from both functions, mapping bytes to key in `p2pk_public_key`

Reply:
I like the idea of moving it to another function, but considering that is_p2pk already checks the length and is used in the new function, do we really need this? is_p2pk is pretty well-tested atm. Just trying to find the reason behind needing to double up on the logic other than just double checking

Reply:
Oof, I'm embarassed that I didn't catch this one

Reply:
Also is_p2pk doesn't do any checking with the bytes of the key, it just makes sure that the op_codes are there and are ordered correctly with the right codes (for compressed/uncompressed keys) within the script. if the key isn't the right size, its not a public key that we can read and is probably malformed therefore it will return None

Reply:
> and is used in the new function

My idea is to do it the other way around - implement the logic in `p2pk_pubkey_bytes` then call `p2pk_pubkey_bytes().is_some()` in `is_p2pk` and `p2pk_pubkey_bytes().map(PublicKey::from_slice).flat_map(Result::ok)` in `p2pk_public_key`

Reply:
Ok, this makes sense. New changes incoming shortly

Reply:
I'm not sure if we can remove the `match self.len()` from is_p2pk because it also verifies that `OP_CHECKSIG` and `OP_PUSHBYTES_{len}` is in the right spot. This means that pubkey bytes still needs to check the length and return the correct bytes if the key is compressed or not. The branching and checks isn't too terrible computationally so this should be ok to have it match twice just to fit 2 cases?

Reply:
> it also verifies that `OP_CHECKSIG` and `OP_PUSHBYTES_{len}` is in the right spot

This verification must be moved to `p2pk_pubkey_bytes` of course

> pubkey bytes still needs to check the length and return the correct bytes if the key is compressed or not

Indeed.

> isn't too terrible computationally so this should be ok to have it match twice

We're not concerned about computation, the compiler will deal with it. We want to have single code to avoid bugs.

Reply:
> Checking of opcodes needs to be moved too - see my comment above.

So the only logic that needs to be in `is_p2pk` is `p2pk_pubkey_bytes.is_some()` because the length checking and op code checking will be moved to `p2pk_pubkey_bytes`? I think should come up with a different name for that function but I can't think of anything good. Maybe `p2pk_pubkey_if_valid`?? something like that I guess